### PR TITLE
Increase Pyxis test coverage

### PR DESCRIPTION
**Pull request by bcrochet**
* Increase the Pyxis test coverage
* Fix some inconsistencies with the Pyxis test harness
* Fix some log messages that were poorly formatted
* Change some magic number to use constants from http package

Signed-off-by: Brad P. Crochet <brad@redhat.com>

### CODE REVIEW

File: certification/pyxis/pyxis.go [hunk 1]
```diff
@@ -60,7 +60,7 @@ func (p *pyxisClient) createImage(ctx context.Context, certImage *CertImage) (*C
 
 	resp, err := p.Client.Do(req)
 	if err != nil {
-		log.Error(err)
+		log.Error(fmt.Errorf("%w: cannot create image in createImage", err))
```
Comment (komish) [inline certification/pyxis/pyxis.go:63]: nit: `cannot create image in createImage` --> `cannot create image` for the log.Error log level? Seems redundant otherwise.

File: certification/pyxis/pyxis_suite_test.go [hunk 2]
```diff
@@ -77,13 +81,26 @@ func (p *pyxisImageHandler) ServeHTTP(response http.ResponseWriter, request *htt
...
+	case request.Method == http.MethodGet && request.Header["X-Api-Key"][0] == "my-bad-401-image-api-token":
+		log.Trace("get with 401")
+		response.WriteHeader(http.StatusUnauthorized)
 	case request.Header["X-Api-Key"][0] == "my-bad-image-api-token":
-		response.WriteHeader(401)
+		response.WriteHeader(http.StatusUnauthorized)
+	case request.Method == http.MethodPost && request.Header["X-Api-Key"][0] == "my-bad-500-image-api-token":
+		response.WriteHeader(http.StatusInternalServerError)
+	case request.Method == http.MethodPost:
+		mustWrite(response, responseString)
```
Comment (komish) [inline certification/pyxis/pyxis_suite_test.go:101]: This looks like it intends to test a variety of scenarios. I'd benefit from a few words in a comment going over the matrix.

File: certification/pyxis/pyxis_suite_test.go [hunk 3]
```diff
@@ -95,11 +112,15 @@ func (p *pyxisRPMManifestHandler) ServeHTTP(response http.ResponseWriter, reques
...
+	case request.Method == http.MethodPost && request.Header["X-Api-Key"][0] == "my-bad-rpmmanifest-409-api-token":
+		response.WriteHeader(http.StatusConflict)
 		mustWrite(response, `{"_id":"foo"}`)
+	case request.Method == http.MethodPost && request.Header["X-Api-Key"][0] == "my-bad-rpmmanifest-401-api-token":
+		response.WriteHeader(http.StatusConflict)
+	case request.Method == http.MethodGet && request.Header["X-Api-Key"][0] == "my-bad-rpmmanifest-401-api-token":
+		response.WriteHeader(http.StatusUnauthorized)
 	case request.Header["X-Api-Key"][0] == "my-bad-rpmmanifest-api-token":
-		response.WriteHeader(401)
+		response.WriteHeader(http.StatusUnauthorized)
```
Comment (komish) [inline certification/pyxis/pyxis_suite_test.go:123]: Same thing here, re: comments

File: certification/pyxis/submit.go [hunk 4]
```diff
@@ -51,31 +51,30 @@ func (p *pyxisClient) SubmitResults(ctx context.Context, certInput *certificatio
...
 		if err != nil {
-			log.Error(err, "could not get image")
+			log.Error(fmt.Errorf("%w: could not get image", err))
 			return nil, err
 		}
 	}
+	if err != nil && err != errors.ErrPyxis409StatusCode {
+		log.Error(fmt.Errorf("%w: could not create image", err))
+		return nil, err
+	}
```
Comment (komish) [inline certification/pyxis/submit.go:64]: I know we (and Go) typically recommends keeping the flow as close to the left-most indentation as possible (as in, nest less), but this seems like a prime opportunity where nesting would fit. 

```Go
if err != nil {
    if err == errors.ErrPyxis409StatusCode { // or a case statement if we want to match more than one type of error
        ...
    }

    log...
    return...
}

File: certification/pyxis/submit.go [hunk 5]
```diff
@@ -51,31 +51,30 @@ func (p *pyxisClient) SubmitResults(ctx context.Context, certInput *certificatio
...
 		return nil, err
 	}
 	if err != nil && err == errors.ErrPyxis409StatusCode {
 		_, err = p.getRPMManifest(ctx, rpmManifest.ImageID)
 		if err != nil {
-			log.Error(err, "could not get rpm manifest")
+			log.Error(fmt.Errorf("%w: could not get rpm manifest", err))
 			return nil, err
 		}
 	}
```
Comment (komish) [inline certification/pyxis/submit.go:80]: This might also benefit from the previous comment on similar code, though I know this didn't really change here.

File: certification/pyxis/submit.go [hunk 6]
```diff
@@ -85,7 +84,7 @@ func (p *pyxisClient) SubmitResults(ctx context.Context, certInput *certificatio
 	for _, artifact := range artifacts {
 		artifact.ImageID = certImage.ID
 		if _, err := p.createArtifact(ctx, &artifact); err != nil {
-			log.Errorf("%s: could not create artifact: %s", err, artifact.Filename)
+			log.Error(fmt.Errorf("%w: could not create artifact: %s", err, artifact.Filename))
```
Comment (komish) [inline certification/pyxis/submit.go:87]: great catch

File: certification/pyxis/submit_test.go [hunk 7]
```diff
@@ -14,264 +14,210 @@ var _ = Describe("Pyxis Submit", func() {
...
+	var certInput certificationInput
 	mux := http.NewServeMux()
 	mux.Handle("/api/v1/projects/certification/id/", &pyxisProjectHandler{})
 	mux.Handle("/api/v1/images", &pyxisImageHandler{})
+	mux.Handle("/api/v1/projects/certification/id/my-image-project-id/images", &pyxisImageHandler{})
 	mux.Handle("/api/v1/images/id/blah/", &pyxisRPMManifestHandler{})
 	mux.Handle("/api/v1/projects/certification/id/my-awesome-project-id/test-results", &pyxisTestResultsHandler{})
 
+	BeforeEach(func() {
+		pyxisClient = NewPyxisClient("my.pyxis.host/api", "my-spiffy-api-token", "my-awesome-project-id", &http.Client{Transport: localRoundTripper{handler: mux}})
```
Comment (komish) [inline certification/pyxis/submit_test.go:26]: nit: multiline this?

File: certification/pyxis/pyxis.go (hunk 1 above)
Comment (bcrochet) [reply certification/pyxis/pyxis.go]: Fixed.

File: certification/pyxis/submit.go (hunk 4 above)
Comment (bcrochet) [reply certification/pyxis/submit.go:78]: This change actually wasn't supposed to be here. :) Only the log fix. I had re-ordered it. I'll play with it.

File: certification/pyxis/pyxis_suite_test.go (hunk 2 above)
Comment (bcrochet) [reply certification/pyxis/pyxis_suite_test.go:103]: How do you feel about nested switch? I think it might help the clarity here. I'm thinking switching on request.Method as the top-level.

File: certification/pyxis/submit_test.go [hunk 8]
```diff
@@ -14,54 +14,83 @@ var _ = Describe("Pyxis Submit", func() {
...
+			})
+		})
+		Context("and the client sends a bad token", func() {
+			JustBeforeEach(func() {
+				pyxisClient.ApiToken = "my-bad-project-api-token"
+			})
+			It("should get an unauthorized", func() {
+				certResults, err := pyxisClient.SubmitResults(ctx, &certInput)
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(MatchError(fmt.Errorf("%w: %s", errors.New("error calling remote API"), "could not retrieve project")))
```
Comment (komish) [inline certification/pyxis/submit_test.go:84]: I mentioned this in Adam's PR as well, but the MatchError feels too granular in that we have the entire error definition written here, and I'm assuming, in the calling code as well.

Assuming these errors are our errors (and not errors coming from the remote API), should we be extracting these into local/package error variables? I don't think they need to go to the top level errors package, but wouldn't it be better to have the string definition of this error in one place? 

I'm not personally offended by checking that an error occurred without asserting the value of it... but I can see it either way. We're going through the effort to wrap, so if you want to test that we get the proper value, that's fine. 

Just my 2c. Feel free to disregard.

File: certification/pyxis/pyxis_suite_test.go (hunk 2 above)
Comment (komish) [reply certification/pyxis/pyxis_suite_test.go:103]: I'm on the fence. But I'm also somewhat on the fence with the general need for a switch here.

I kind of get the need to do multiple scenarios. I think what I'm wondering is, would there be a way to restructure the test that would require less switching?

**Review by komish — CHANGES_REQUESTED** [review]
Couple of comments in line. Only submitting a review again because GH says my comments are pending.

File: certification/pyxis/pyxis_suite_test.go (hunk 2 above)
Comment (bcrochet) [reply certification/pyxis/pyxis_suite_test.go:103]: I take it back. I think I want to stick with a flat switch. Then I don't end up making duplicates for ones that are the same for Get and Post.

File: certification/pyxis/pyxis_suite_test.go (hunk 3 above)
Comment (bcrochet) [reply certification/pyxis/pyxis_suite_test.go:125]: I'm making a "top-level" comment about these.

File: certification/pyxis/pyxis_suite_test.go (hunk 2 above)
Comment (acornett21) [reply certification/pyxis/pyxis_suite_test.go:108]: FWIW, when I was adding tests to this, I thought about a nested switch as well when I thought I might need to marshal the request payload for an edge case, but when I coded that I hated how it read. I ended up just going with an `if` for the one case, and default return for everything else. Hopefully we never have a case were need more then two options within a given switch (ie GET vs POST).

File: certification/pyxis/submit_test.go (hunk 8 above)
Comment (bcrochet) [reply certification/pyxis/submit_test.go:91]: I'm not a fan of matching the error either. I'm just going to remove.

**Comment by bcrochet** [comment]
@komish I'm not sure why it won't let me respond directly... Anyway, about the switch in the handlers. Basically, I'm trying to minimize the number of handlers. Otherwise, I think we'd end up with a bunch of small handlers. I do admit that it's rather fragile. If you have any suggestions for improving this mocked up http server, I'm all ears.

File: certification/pyxis/submit_test.go (hunk 8 above)
Comment (acornett21) [reply certification/pyxis/submit_test.go]: Why don't you all like matching the error? To me it makes sense to match the error so we know that the code is covered and hitting the proper flow we expect. We match values on `happy` path testing, ie `registry = docker.io` as an example to make sure we did something properly. I really don't understand not wanting to do the same for exceptions.

Comment (komish) [reply certification/pyxis/submit_test.go]: @acornett21 for me, it's because the message we're testing for is a message that we set. I think it would be one thing if we were testing the external systems error messages (as in, make sure _this code path_ results in _that error_), but I don't think that's what's happening here.

**Comment by komish** [comment]
> @komish I'm not sure why it won't let me respond directly... Anyway, about the switch in the handlers. Basically, I'm trying to minimize the number of handlers. Otherwise, I think we'd end up with a bunch of small handlers. I do admit that it's rather fragile. If you have any suggestions for improving this mocked up http server, I'm all ears.

I think I need to spend some time digging further into HTTP client/server testing in Go to really be able to offer an alternative. Which is to say that I don't have one, and I may as well just find myself reaching the same conclusion/implementation you have here after I spend time with it.

to that end, I think I'm comfortable with the way it's being tested even though it does somewhat feel/read a bit wild. The amount of running code (vs. test code) being touched is also minimal. Overall: 👍

**Review by komish — APPROVED** [review]
/lgtm

**Review by acornett21 — APPROVED** [review]
